### ISSUE TITLE
Is it still a good design if all attributes of a person are not immutable?

### ISSUE THREAD

**Issue by agarwal-anjali** [nus-cs2103-AY2223S1/forum#361]
Initially the edit command was only to edit an attribute of a single person at a time and when the execute method of the edit command was called a new instance of  a Person was created with the edited attributes and the person list in the address book was updated.

But now, if I have a feature in which I am editing an attribute of all persons (say attribute X) in the address book, is it better to have attribute X mutable so that when the execute method of the edit command for this attribute is called we can modify the value for this attribute for all the persons in the address book or should we maintain the immutability of the attributes of a person and create a new instance of each person for the edited attribute X (equivalent to replicating the address book in the memory with the updated attribute) every time attribute X  is being edited?

**Comment by damithc** [comment]
@agarwal-anjali This is a good question for a discussion. Have you already tried either of the options yet?

**Comment by clarence-chew** [comment]
One reason to prefer an immutable person is that it lessens the memory impact of the implementation of the undo command using a VersionedAddressBook. In this case, the same Person objects can be used across multiple VersionedAddressBooks.

**Comment by agarwal-anjali** [comment]
> @agarwal-anjali This is a good question for a discussion. Have you already tried either of the options yet?

Yes, in my current implementation the attribute X is mutable for a person!

**Comment by agarwal-anjali** [comment]
> One reason to prefer an immutable person is that it lessens the memory impact of the implementation of the undo command using a VersionedAddressBook. In this case, the same Person objects can be used across multiple VersionedAddressBooks.

Yeah thank you, it makes sense!
I shall consider making all the attributes of a person immutable.

**Comment by damithc** [comment]
Unless the app is expected to store a large amount of data, memory usage is unlikely to be a high-priority concern. If the current code works and there is no clear evidence of a gain from using an alternative approach, there is no good reason to change. c/f https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-premature-optimizations 

A related thread from a previous semester https://github.com/nus-cs2103-AY2122S2/forum/issues/193
